File: src/main/kotlin/com/natu/ftax/ledger/infrastructure/LedgerBookEntity.java
package com.natu.ftax.ledger.infrastructure;

import com.natu.ftax.ledger.domain.LedgerBook;
import jakarta.persistence.*;

import java.util.List;
import java.util.stream.Collectors;

@Entity
@Table(name = "ledger_books")
public class LedgerBookEntity {

    @Id
    private String id;

    @OneToMany(mappedBy = "ledgerBook", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    private List<LedgerEntryEntity> ledgerEntries;

    protected LedgerBookEntity() {
        // No-arg constructor required by JPA
    }

    public LedgerBookEntity(String id, List<LedgerEntryEntity> ledgerEntries) {
        this.id = id;
        this.ledgerEntries = ledgerEntries;
    }

    public static LedgerBookEntity fromDomain(LedgerBook ledgerBook) {
        return new LedgerBookEntity(
            ledgerBook.getId(),
            ledgerBook.getLedgerEntries().stream().map(entry -> LedgerEntryEntity.fromDomain(entry, ledgerBook.getId())).collect(Collectors.toList())
        );
    }

    public LedgerBook toDomain() {
        return new LedgerBook(
            id,
            ledgerEntries.stream().map(LedgerEntryEntity::toDomain).collect(Collectors.toList())
        );
    }
}

File: src/main/kotlin/com/natu/ftax/ledger/infrastructure/LedgerBookEntity.kt
package com.natu.ftax.ledger.infrastructure

import com.natu.ftax.ledger.domain.LedgerBook
import jakarta.persistence.*

@Entity
@Table(name = "Ledger_Books")
class LedgerBookEntity(
    @Id
    val id: String,

    @OneToMany(mappedBy = "ledgerBook", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    val ledgerEntries: MutableList<LedgerEntryEntity> = mutableListOf()
) {
    // No-arg constructor required by JPA
    protected constructor() : this(id = "dum", ledgerEntries = mutableListOf())

    companion object {
        fun fromDomain(ledgerBook: LedgerBook): LedgerBookEntity {
            return LedgerBookEntity(
                id = ledgerBook.id,
                ledgerEntries = ledgerBook.ledgerEntries.map { LedgerEntryEntity.fromDomain(it, ledgerBook.id) }
                    .toMutableList()

            )
        }
    }

    fun toDomain(): LedgerBook {
        return LedgerBook(
            id = id,
            ledgerEntries = ledgerEntries.map { it.toDomain() }.toMutableList()
        )
    }
}

